File: rust/equity-var-engine/tests/test_monte_carlo.rs
//! Monte Carlo VaR: convergence to the parametric closed form within 3 SE,
//! Student-t tail fattening, bitwise seed determinism, moment matching.

use eq_var_engine::expected_shortfall::normal_es;
use eq_var_engine::matrix::{covariance_from_vols, sample_covariance, Matrix};
use eq_var_engine::monte_carlo::*;
use eq_var_engine::parametric::{parametric_var, portfolio_sigma};
use eq_var_engine::rng::Rng;
use eq_var_engine::stats::normal_ppf;
use eq_var_engine::TailModel;

fn demo_cov() -> Matrix {
    let vols = [0.010, 0.015, 0.020];
    let corr = Matrix::from_vec(3, 3, vec![1.0, 0.5, 0.25, 0.5, 1.0, 0.3, 0.25, 0.3, 1.0]).unwrap();
    covariance_from_vols(&vols, &corr).unwrap()
}

const EXPOSURES: [f64; 3] = [1.0e6, -5.0e5, 2.0e5];

#[test]
fn bitwise_seed_determinism() {
    let cov = demo_cov();
    let a = monte_carlo_var(&EXPOSURES, &cov, 0.01, 20_000, TailModel::Normal, 42).unwrap();
    let b = monte_carlo_var(&EXPOSURES, &cov, 0.01, 20_000, TailModel::Normal, 42).unwrap();
    assert_eq!(a, b); // bitwise, not approximate
    let ea = monte_carlo_es(&EXPOSURES, &cov, 0.01, 20_000, TailModel::Normal, 42).unwrap();
    let eb = monte_carlo_es(&EXPOSURES, &cov, 0.01, 20_000, TailModel::Normal, 42).unwrap();
    assert_eq!(ea, eb);
    let c = monte_carlo_var(&EXPOSURES, &cov, 0.01, 20_000, TailModel::Normal, 43).unwrap();
    assert_ne!(a, c); // a different seed must move the estimate
}

#[test]
fn normal_converges_to_parametric_within_3se() {
    let cov = demo_cov();
    let exact_var = parametric_var(&EXPOSURES, &cov, 0.01, TailModel::Normal).unwrap();
    let exact_es = normal_es(portfolio_sigma(&EXPOSURES, &cov).unwrap(), 0.01, 0.0).unwrap();
    let pnl = monte_carlo_pnl(&EXPOSURES, &cov, 200_000, TailModel::Normal, 7).unwrap();
    let mc_var = monte_carlo_var(&EXPOSURES, &cov, 0.01, 200_000, TailModel::Normal, 7).unwrap();
    let mc_es = monte_carlo_es(&EXPOSURES, &cov, 0.01, 200_000, TailModel::Normal, 7).unwrap();
    let se = var_order_statistic_se(&pnl, 0.01).unwrap();
    assert!(se > 0.0);
    assert!(se < 0.02 * exact_var);
    assert!((mc_var - exact_var).abs() < 3.0 * se, "mc={mc_var} exact={exact_var} se={se}");
    assert!((mc_es - exact_es).abs() < 4.0 * se);
}

#[test]
fn student_t_fatter_than_normal_at_99_and_converges_to_closed_form() {
    let cov = demo_cov();
    let n_var = monte_carlo_var(&EXPOSURES, &cov, 0.01, 200_000, TailModel::Normal, 11).unwrap();
    let n_es = monte_carlo_es(&EXPOSURES, &cov, 0.01, 200_000, TailModel::Normal, 11).unwrap();
    let t_var = monte_carlo_var(
        &EXPOSURES,
        &cov,
        0.01,
        200_000,
        TailModel::StudentT { df: 5.0 },
        11,
    )
    .unwrap();
    let t_es = monte_carlo_es(
        &EXPOSURES,
        &cov,
        0.01,
        200_000,
        TailModel::StudentT { df: 5.0 },
        11,
    )
    .unwrap();
    assert!(t_var > n_var);
    assert!(t_es > n_es);
    let exact_t = parametric_var(&EXPOSURES, &cov, 0.01, TailModel::StudentT { df: 5.0 }).unwrap();
    let pnl_t = monte_carlo_pnl(&EXPOSURES, &cov, 200_000, TailModel::StudentT { df: 5.0 }, 11)
        .unwrap();
    let se_t = var_order_statistic_se(&pnl_t, 0.01).unwrap();
    assert!((t_var - exact_t).abs() < 3.0 * se_t);
}

#[test]
fn es_dominates_var() {
    let cov = demo_cov();
    for seed in [0u64, 1, 2] {
        let var = monte_carlo_var(&EXPOSURES, &cov, 0.025, 50_000, TailModel::Normal, seed).unwrap();
        let es = monte_carlo_es(&EXPOSURES, &cov, 0.025, 50_000, TailModel::Normal, seed).unwrap();
        assert!(es >= var, "seed={seed}");
    }
}

#[test]
fn simulate_factor_returns_matches_target_covariance_normal_and_t() {
    let cov = demo_cov();
    for tail in [TailModel::Normal, TailModel::StudentT { df: 6.0 }] {
        let scen = simulate_factor_returns(&cov, 200_000, tail, 3).unwrap();
        let hat = sample_covariance(&scen).unwrap();
        for i in 0..3 {
            assert!((hat.get(i, i) / cov.get(i, i) - 1.0).abs() < 0.05, "i={i}");
            for j in 0..3 {
                let tol = 0.10 * (cov.get(i, i) * cov.get(j, j)).sqrt();
                assert!((hat.get(i, j) - cov.get(i, j)).abs() < tol);
            }
        }
    }
}

#[test]
fn rng_uniform_in_open_interval_and_gaussian_moments() {
    let mut rng = Rng::new(123);
    let (mut s, mut s2) = (0.0, 0.0);
    for _ in 0..100_000 {
        let u = rng.uniform();
        assert!(u > 0.0 && u < 1.0);
        let g = normal_ppf(u).unwrap();
        s += g;
        s2 += g * g;
    }
    assert!((s / 1e5).abs() < 0.015);
    assert!((s2 / 1e5 - 1.0).abs() < 0.02);
}

#[test]
fn rng_chi_squared_mean_and_variance() {
    let mut rng = Rng::new(99);
    let df = 6.0;
    let (mut s, mut s2) = (0.0, 0.0);
    let n = 50_000;
    for _ in 0..n {
        let x = rng.chi_square(df);
        assert!(x > 0.0);
        s += x;
        s2 += x * x;
    }
    let m = s / n as f64;
    assert!((m - df).abs() < 0.07);
    assert!((s2 / n as f64 - m * m - 2.0 * df).abs() < 0.5);
}

#[test]
fn monte_carlo_validation() {
    let cov = demo_cov();
    assert!(monte_carlo_var(&[], &cov, 0.01, 1000, TailModel::Normal, 0).is_err());
    assert!(monte_carlo_var(&[1.0], &cov, 0.01, 1000, TailModel::Normal, 0).is_err());
    assert!(monte_carlo_var(&EXPOSURES, &cov, 0.6, 1000, TailModel::Normal, 0).is_err());
    assert!(monte_carlo_var(&EXPOSURES, &cov, 0.01, 1000, TailModel::StudentT { df: 2.0 }, 0)
        .is_err());
    assert!(simulate_factor_returns(&cov, 0, TailModel::Normal, 0).is_err());
    let tiny_pnl = vec![0.0; 50];
    assert!(var_order_statistic_se(&tiny_pnl, 0.01).is_err());
}

File: cpp/equity-var-engine/tests/test_monte_carlo.cpp
// Monte Carlo VaR: convergence to the parametric closed form within 3 SE,
// Student-t tail fattening, bitwise seed determinism, moment matching.
#include <gtest/gtest.h>

#include <cmath>
#include <vector>

#include "eqvar/expected_shortfall.hpp"
#include "eqvar/monte_carlo.hpp"
#include "eqvar/parametric.hpp"
#include "eqvar/returns.hpp"
#include "eqvar/stats.hpp"

using namespace eqvar;

namespace {

Matrix demo_cov() {
    const std::vector<double> vols = {0.010, 0.015, 0.020};
    const Matrix corr(3, 3, {1.0, 0.5, 0.25, 0.5, 1.0, 0.3, 0.25, 0.3, 1.0});
    return covariance_from_vols(vols, corr);
}

const std::vector<double> kExposures = {1.0e6, -5.0e5, 2.0e5};

}  // namespace

TEST(MonteCarlo, BitwiseSeedDeterminism) {
    const Matrix cov = demo_cov();
    const MonteCarloResult a = monte_carlo_var(kExposures, cov, 0.01, 20'000, Dist::Normal, 6.0, 42);
    const MonteCarloResult b = monte_carlo_var(kExposures, cov, 0.01, 20'000, Dist::Normal, 6.0, 42);
    EXPECT_EQ(a.var, b.var);  // bitwise, not approximate
    EXPECT_EQ(a.es, b.es);
    EXPECT_EQ(a.var_se, b.var_se);
    const MonteCarloResult c = monte_carlo_var(kExposures, cov, 0.01, 20'000, Dist::Normal, 6.0, 43);
    EXPECT_NE(a.var, c.var);  // a different seed must move the estimate
}

TEST(MonteCarlo, NormalConvergesToParametricWithin3SE) {
    const Matrix cov = demo_cov();
    const double exact_var = parametric_var(kExposures, cov, 0.01);
    const double exact_es = normal_es(portfolio_sigma(kExposures, cov), 0.01);
    const MonteCarloResult mc =
        monte_carlo_var(kExposures, cov, 0.01, 200'000, Dist::Normal, 6.0, 7);
    ASSERT_GT(mc.var_se, 0.0);
    EXPECT_LT(mc.var_se, 0.02 * exact_var);  // 200k paths: SE well under 2 %
    EXPECT_NEAR(mc.var, exact_var, 3.0 * mc.var_se);
    // ES averages the tail; allow a slightly wider band on the same SE scale.
    EXPECT_NEAR(mc.es, exact_es, 4.0 * mc.var_se);
}

TEST(MonteCarlo, StudentTFatterThanNormalAt99AndConvergesToClosedForm) {
    const Matrix cov = demo_cov();
    const MonteCarloResult n =
        monte_carlo_var(kExposures, cov, 0.01, 200'000, Dist::Normal, 6.0, 11);
    const MonteCarloResult t =
        monte_carlo_var(kExposures, cov, 0.01, 200'000, Dist::StudentT, 5.0, 11);
    EXPECT_GT(t.var, n.var);
    EXPECT_GT(t.es, n.es);
    const double exact_t = parametric_var(kExposures, cov, 0.01, Dist::StudentT, 5.0);
    EXPECT_NEAR(t.var, exact_t, 3.0 * t.var_se);
}

TEST(MonteCarlo, EsDominatesVar) {
    const Matrix cov = demo_cov();
    for (std::uint64_t seed : {0ULL, 1ULL, 2ULL}) {
        const MonteCarloResult r =
            monte_carlo_var(kExposures, cov, 0.025, 50'000, Dist::Normal, 6.0, seed);
        EXPECT_GE(r.es, r.var) << "seed=" << seed;
    }
}

TEST(SimulateFactorReturns, MatchesTargetCovarianceNormalAndT) {
    const Matrix cov = demo_cov();
    for (Dist dist : {Dist::Normal, Dist::StudentT}) {
        const Matrix scen = simulate_factor_returns(cov, 200'000, dist, 6.0, 3);
        const Matrix hat = sample_covariance(scen);
        // Variances within 5 % relative, covariances within 10 %: ~ 3-4 MC SE
        // at 200k paths (t moments converge slower — same target covariance).
        for (std::size_t i = 0; i < 3; ++i) {
            EXPECT_NEAR(hat(i, i) / cov(i, i), 1.0, 0.05) << "i=" << i;
            for (std::size_t j = 0; j < 3; ++j) {
                EXPECT_NEAR(hat(i, j), cov(i, j), 0.10 * std::sqrt(cov(i, i) * cov(j, j)));
            }
        }
    }
}

TEST(RandomStreamTest, UniformInOpenIntervalAndGaussianMoments) {
    RandomStream rng(123);
    double s = 0.0, s2 = 0.0;
    for (int i = 0; i < 100'000; ++i) {
        const double u = rng.uniform();
        EXPECT_GT(u, 0.0);
        EXPECT_LT(u, 1.0);
        const double g = normal_ppf(u);
        s += g;
        s2 += g * g;
    }
    EXPECT_NEAR(s / 1e5, 0.0, 0.015);   // ~ 4.7 SE
    EXPECT_NEAR(s2 / 1e5, 1.0, 0.02);
}

TEST(RandomStreamTest, ChiSquaredMeanAndVariance) {
    RandomStream rng(99);
    const double df = 6.0;
    double s = 0.0, s2 = 0.0;
    const int n = 50'000;
    for (int i = 0; i < n; ++i) {
        const double x = rng.chi_squared(df);
        EXPECT_GT(x, 0.0);
        s += x;
        s2 += x * x;
    }
    const double m = s / n;
    EXPECT_NEAR(m, df, 0.07);                       // E = df, SE ~ 0.015
    EXPECT_NEAR(s2 / n - m * m, 2.0 * df, 0.5);     // Var = 2 df
}

TEST(MonteCarlo, BootstrapSeSameSeedIsBitwiseIdentical) {
    const Matrix cov = demo_cov();
    const Matrix scen = simulate_factor_returns(cov, 20'000, Dist::Normal, 6.0, 5);
    const std::vector<double> pnl = portfolio_pnl(scen, kExposures);
    const double se_a = mc_bootstrap_se(pnl, 0.01, 200, 7);
    const double se_b = mc_bootstrap_se(pnl, 0.01, 200, 7);
    EXPECT_EQ(se_a, se_b);  // bitwise, not approximate
    const double se_c = mc_bootstrap_se(pnl, 0.01, 200, 8);
    EXPECT_NE(se_a, se_c);  // a different seed must move the estimate
}

TEST(MonteCarlo, BootstrapSeThrowsOnFewerThan10Scenarios) {
    const std::vector<double> tiny(9, -1.0);
    EXPECT_THROW(mc_bootstrap_se(tiny, 0.01), std::invalid_argument);
    const std::vector<double> ok(10, -1.0);
    EXPECT_NO_THROW(mc_bootstrap_se(ok, 0.01, 50));
}

TEST(MonteCarlo, BootstrapSeThrowsOnAlphaOutsideValidRange) {
    const std::vector<double> pnl(100, -1.0);
    EXPECT_THROW(mc_bootstrap_se(pnl, 0.0), std::invalid_argument);
    EXPECT_THROW(mc_bootstrap_se(pnl, 0.5), std::invalid_argument);
    EXPECT_THROW(mc_bootstrap_se(pnl, -0.1), std::invalid_argument);
    EXPECT_THROW(mc_bootstrap_se(pnl, 0.6), std::invalid_argument);
}

TEST(MonteCarlo, BootstrapSeWithinSaneMultipleOfOrderStatisticSe) {
    const Matrix cov = demo_cov();
    const Matrix scen = simulate_factor_returns(cov, 50'000, Dist::Normal, 6.0, 21);
    const std::vector<double> pnl = portfolio_pnl(scen, kExposures);
    const MonteCarloResult ref = mc_tail_metrics(pnl, 0.01);
    const double boot_se = mc_bootstrap_se(pnl, 0.01, 300, 3);
    ASSERT_GT(ref.var_se, 0.0);
    ASSERT_GT(boot_se, 0.0);
    // Different estimators, not expected to agree closely; just guard
    // against wild divergence (the bootstrap is known to run a bit higher
    // in deep tails, per the order-statistic estimator's low bias there).
    EXPECT_LT(boot_se, 3.0 * ref.var_se);
    EXPECT_GT(boot_se, ref.var_se / 3.0);
}

TEST(MonteCarlo, BootstrapSeHandlesDegenerateZeroVarianceSampleWithoutNaN) {
    const std::vector<double> flat(200, 42.0);
    const double se = mc_bootstrap_se(flat, 0.01, 100, 1);
    EXPECT_TRUE(std::isfinite(se));
    EXPECT_DOUBLE_EQ(se, 0.0);
}

TEST(MonteCarlo, Validation) {
    const Matrix cov = demo_cov();
    EXPECT_THROW(monte_carlo_var(std::vector<double>{}, cov, 0.01), std::invalid_argument);
    EXPECT_THROW(monte_carlo_var(std::vector<double>{1.0}, cov, 0.01), std::invalid_argument);
    EXPECT_THROW(monte_carlo_var(kExposures, cov, 0.6), std::invalid_argument);
    EXPECT_THROW(monte_carlo_var(kExposures, cov, 0.01, 1000, Dist::StudentT, 2.0),
                 std::invalid_argument);
    EXPECT_THROW(simulate_factor_returns(cov, 0), std::invalid_argument);
    const std::vector<double> tiny(50, 0.0);
    EXPECT_THROW(mc_tail_metrics(tiny, 0.01), std::invalid_argument);
}
